### ISSUE TITLE
SRT input support

### ISSUE THREAD

**Issue by nemzo89** [AirenSoft/OvenMediaEngine#80]
SRT is already being used between the origin and edge servers, but it would be interesting to have input support as well. Are SRT input support planned for v1.0? OBS Studio are planning to support SRT output as of version 25.0, and the code is already supported in the release candidates.

So far I can see rtmp and rtsp support in the src/projects/providers/ folder of the development branch.

Thank you so much for providing the ovenmediaengine server! I spent days looking for optimal solutions for webrtc livestreaming with sub second latencies, and ovenmediaengine is far ahead of even very expensive market leaders. You guys are a blessing, and it is a joy to see it being actively developed.

**Comment by getroot** [comment]
Thank you for the good information and compliments. 

We think OBS is one of the most important Encoder to OvenMediaEngine. So if OBS supports SRT Output we will support SRT Input.

**Comment by vixns** [comment]
https://obsproject.com/forum/threads/obs-studio-25-0-release-candidate.116067/

**Comment by dimiden** [comment]
@vixns
Thanks for the information. I'll review the SRT input with the OME development team soon.

**Comment by getroot** [comment]
We are going to support this feature.

**Comment by ynerant** [comment]
Hi!
I'm glad to see that you're working on the SRT support!
I'm reading the docs about the MPEG-2-TS input, and I see that there is only one stream per port, maybe because of the specification of the transport that does not allow to specify a stream key/id.
But fortunately SRT allow to enter a stream id, with the `streamid` parameter :)
It can be very nice if you can handle this parameter properly :) Maybe is it already planned?

**Comment by belthesar** [comment]
Heya! Curious to see where this is on the roadmap. I'm interested to see if OME can provide the backbone for our streaming platform framework, but I would _really_ like to use SRT in lieu of RTMP for ingest. The durability concerns of WebRTC are currently unattractive for me, so having something more reliable for ingest, but significantly faster than RTMP is ideal. Thanks!

**Comment by getroot** [comment]
Hi. Unfortunately this feature is still in the backlog. We've been focusing more on streaming than Ingest recently. We have found that several people are interested in this feature, so we will discuss the priorities for this feature.

**Comment by creisner** [comment]
Any update on this topics. I'm also looking for SRT input.

**Comment by medcelerate** [comment]
Looking forward to this feature. SRT would be incredible!

**Comment by reustaquiojr** [comment]
Looking for SRT input too !! Would be a nice feature !

**Comment by getroot** [comment]
I just started studying about the SRT specification of OBS.

https://obsproject.com/wiki/Streaming-With-SRT-Protocol

Referring to the above document, I have figured out that the MEPGTS format is transmitted by SRT transport. But I don't know where to find specifications like the streamid parameter mentioned by @ynerant 

Can someone please tell me where is the SRT (MPEGTS) specification that OBS (or ffmpeg) is using?

**Comment by medcelerate** [comment]
It’s all in the haivision repository on github, they have a reference
implementation.

**Comment by getroot** [comment]
@medcelerate 
Thank you! 

https://github.com/Haivision/srt/blob/master/docs/features/access-control.md

**Comment by getroot** [comment]
I implemented the SRT Provider and committed it to the Master branch. If someone can help me test this feature, I'd appreciate it.

![image](https://user-images.githubusercontent.com/17779085/116431081-65ad4080-a882-11eb-9714-66dbdb3e123f.png)

You can use the SRT Provider by following the procedure below.

**[Add SRT related setting to Server.xml]**
```
<Bind>
	<Providers>
		<SRT>
			<Port>9999</Port>
		</SRT>
...

<VirtualHosts>
	<VirtualHost>
		<Providers>
			<SRT />
			...
```

**[Set as follows in OBS]**

OME classifies domain/app/stream using SRT's streamid. The specifications of streamid are as follows.

`streamid=srt://host[:port]/app/stream`

You can set OBS as follows.

srt://192.168.0.200:9999?streamid=srt://192.168.0.200:9999/app/stream

![image](https://user-images.githubusercontent.com/17779085/116431060-5f1ec900-a882-11eb-9ab6-907305fa6019.png)

Anyone please do not spare me a lot of bug reports and advice.

**Comment by medcelerate** [comment]
I’ll build and do some testing.

**Comment by bchah** [comment]
Building now! Thanks @getroot 👍🏻👍🏻👍🏻

**Comment by bchah** [comment]
@getroot You should celebrate my friend, because the SRT provider is working well!

![Screen Shot 2021-04-28 at 7 58 40 PM](https://user-images.githubusercontent.com/33604674/116486768-5ddaa580-a85c-11eb-8faf-d06a818a0647.jpg)

**Comment by getroot** [comment]
Today I applied this feature to the default configuration. If you find another issue, please open a new issue. 
Thank you for all of your contributions.

**Comment by sehadnassim** [comment]
Not working for me, log :  
![image](https://user-images.githubusercontent.com/42390703/117241456-e7313f80-ae3b-11eb-92f5-a547169a5f77.png)
Can you tell me what browser are you using, also did you try it with tls  webRTC?
the log above is after doing the exact same as you show, over non-tls webrtc, could you help please ?

**Comment by getroot** [comment]
@sehadnassim 

I have to review your Server.xml, the entire log, and the encoding options of OBS to find out why it won't play.

Does it play well with WebRTC when transmitted by RTMP? Check it out first. Otherwise, there may be a problem with your network settings, etc.

For reference, the last Master version has changed the streamid standard of SRT.
https://airensoft.gitbook.io/ovenmediaengine/v/master/live-source/srt-beta

**Comment by sehadnassim** [comment]
Yes it works with WebRTC when transmitted by RTMP, I also tried with the new streamid standard, but I got this error : 
![image](https://user-images.githubusercontent.com/42390703/117244020-f8307f80-ae40-11eb-9e3f-d0335e0bb3d9.png)

Here is my Server.xml : 
<?xml version="1.0" encoding="UTF-8"?>

<Server version="8">
	<Name>OvenMediaEngine</Name>
	
	<Type>origin</Type>
	
	<IP>*</IP>

	
	<StunServer>stun.l.google.com:19302</StunServer>

	
	<Bind>
		
		

		<Providers>
			<RTMP>
				<Port>1935</Port>
				
			</RTMP>
			<SRT>
				<Port>9999</Port>
				
			</SRT>
			<MPEGTS>
				
				<Port>4000-4004,4005/udp</Port>
			</MPEGTS>
			<WebRTC>
				<Signalling>
					<Port>3333</Port>
					
					 <TLSPort>3334</TLSPort> 
					
				</Signalling>

				<IceCandidates>
					
					<TcpRelay>*:3478</TcpRelay>
					<IceCandidate>*:10000-10005/udp</IceCandidate>

					
				</IceCandidates>
			</WebRTC>
		</Providers>

		<Publishers>
			<OVT>
				<Port>9000</Port>
				
			</OVT>
	 
				
				


		
				
				
		
			<WebRTC>
				<Signalling>
					<Port>3333</Port>
					
					
					
				</Signalling>

				<IceCandidates>
					
					<TcpRelay>*:3478</TcpRelay>
					<IceCandidate>*:10000-10005/udp</IceCandidate>

					
				</IceCandidates>
			</WebRTC>

                       
		</Publishers>
	</Bind>

	
	

	
	

	<VirtualHosts>
		
		<VirtualHost include="VHost*.xml" />
		<VirtualHost>
			<Name>default</Name>

			
			<Host>
				<Names>
					
					<Name>*</Name>
				</Names>
				
				<TLS>
					<CertPath>/home/ubuntu/private/certificate.cer</CertPath>
					<KeyPath>/home/ubuntu/private/private.key</KeyPath>
					<ChainCertPath>/home/ubuntu/private/certificate.cer</ChainCertPath>
				</TLS>
				
			</Host>

			

			
			
			<Applications>
				<Application>
					<Name>app</Name>
					
					<Type>live</Type>
					<OutputProfiles>
						<OutputProfile>
							<Name>bypass_stream</Name>
							<OutputStreamName>${OriginStreamName}</OutputStreamName>
							<Encodes>
								<Audio>
									<Bypass>true</Bypass>
								</Audio>
								<Video>
									<Bypass>true</Bypass>
								</Video>
								<Audio>
									<Codec>opus</Codec>
									<Bitrate>128000</Bitrate>
									<Samplerate>48000</Samplerate>
									<Channel>2</Channel>
								</Audio>
								
							</Encodes>
						</OutputProfile>
					</OutputProfiles>
					<Providers>
						<OVT />
						<WebRTC />
						<RTMP />
						<SRT />
						<MPEGTS>
							<StreamMap>
								
								<Stream>
									<Name>stream_${Port}</Name>
									<Port>4000,4001-4004</Port>
								</Stream>
								<Stream>
									<Name>stream_4005</Name>
									<Port>4005</Port>
								</Stream>
							</StreamMap>
						</MPEGTS>
						<RTSPPull />
						<WebRTC>
							<Timeout>30000</Timeout>
						</WebRTC>
					</Providers>
					<Publishers>
						<SessionLoadBalancingThreadCount>8</SessionLoadBalancingThreadCount>
						<OVT />
						<WebRTC>
							<Timeout>30000</Timeout>
							<Rtx>true</Rtx>
							<Ulpfec>true</Ulpfec>
						</WebRTC>
				
				 
						<LLDASH>
							<SegmentDuration>5</SegmentDuration>
							<CrossDomains>
								<Url>*</Url>
							</CrossDomains>
						</LLDASH>
					</Publishers>
				</Application>
			</Applications>
		</VirtualHost>
	</VirtualHosts>
</Server>

**Comment by getroot** [comment]
@sehadnassim 
Looking at the log, it is not the latest master version. Get the latest master version with git pull.

**Comment by sehadnassim** [comment]
When testing the latest master version i have this problem below, Th path to srt : srt://IP:9999?streamid=srt://IP:9999/app/mosaic

WebRTC : ws://IP:3333/app/mosaic

![image](https://user-images.githubusercontent.com/42390703/117313670-c6460a00-ae8e-11eb-883d-82284d16b0cc.png)

**Comment by getroot** [comment]
Please provide entire log file, not edited capture. and streamid must be encoded by percent encoding.

**Comment by sehadnassim** [comment]
here are logs, over RTMP it works well. streamid : srt://ip:9999?streamid=srt%3A//ip%3A9999/app/mosaic%3Fquery%3Dvalue
[ovenmediaengine.txt](https://github.com/AirenSoft/OvenMediaEngine/files/6436620/ovenmediaengine.txt)

**Comment by getroot** [comment]
@sehadnassim 
The out-of-order packet issue doesn't happen in my environment. Is your OBS up-to-date? It seems that many packets are being lost due to out-of-order (but it may be possible to play back at this time as well).

For the encoding setting of OBS, refer to the following URL.

https://airensoft.gitbook.io/ovenmediaengine/v/master/getting-started#example-of-using-obs-encoder

![image](https://user-images.githubusercontent.com/17779085/117420153-3ff5fa80-af58-11eb-85a7-ca17233a367c.png)
![image](https://user-images.githubusercontent.com/17779085/117420186-45ebdb80-af58-11eb-9484-983d68679935.png)

**Comment by sehadnassim** [comment]
I have the exact same version as yours, now i do not have the out of order error but the client keeps disconnecting.
What i do not understant si that it works great for RTMP so this cannot be a network problem ? 
![image](https://user-images.githubusercontent.com/42390703/117454252-9b22ef80-af4e-11eb-92de-fe469592f722.png)

**Comment by getroot** [comment]
@sehadnassim 
Are the encoding options for OBS the same as what I captured and posted? This seems to be a case where the WebRTC player (do you use chrome?) hangs up because it has not received the keyframe for a certain period of time or cannot decode. I need to send it to your server and play it back to be able to analyze it accurately. Can you provide your srt input address and webrtc play address?

**Comment by sehadnassim** [comment]
> @sehadnassim
> Are the encoding options for OBS the same as what I captured and posted? This seems to be a case where the WebRTC player (do you use chrome?) hangs up because it has not received the keyframe for a certain period of time or cannot decode. I need to send it to your server and play it back to be able to analyze it accurately. Can you provide your srt input address and webrtc play address?

Yes iam using the same encoding options. i tried with chrome, chromium, edge and firefox. 
SRT : srt://195.148.125.88:9999?streamid=srt%3A//195.148.125.88%3A9999/app/mosaic%3Fquery%3Dvalue 
WebRTC : ws://195.148.125.88:3333/app/mosaic
i have tested it through secure webrtc wss://195.148.125.88:3334/app/mosaic

**Comment by getroot** [comment]
![image](https://user-images.githubusercontent.com/17779085/117468979-35575780-af90-11eb-9c03-82845ab80811.png)

I played a stream with WebRTC/tcp on your server. With UDP, no packets are coming into my browser. When I connected via TCP, I could see that it was playing normally. Maybe the UDP port is blocked by a firewall or some other network setup problem exists.

**Comment by sehadnassim** [comment]
> ![image](https://user-images.githubusercontent.com/17779085/117468979-35575780-af90-11eb-9c03-82845ab80811.png)
> 
> I played a stream with WebRTC/tcp on your server. With UDP, no packets are coming into my browser. When I connected via TCP, I could see that it was playing normally. Maybe the UDP port is blocked by a firewall or some other network setup problem exists.

Right it works with TCP, but i dont thin UDP is blocked because it works with RTMP, this is weird. However Thank you for your quick answer and help !
